pallene/checker.lua

```diff
@@ -668,7 +668,14 @@ end
 -- type conversion.
 function FunChecker:check_exp_synthesize(exp)
     local tag = exp._tag
-    if     tag == "ast.Exp.Nil" then
+
+    if exp._type then
+        -- Fallthrough
+        -- Expression already synthesized. This occurs when
+        -- expand_function_returns is called. This ensures the work is done
+        -- only once for call function and call method expressions.
+
```

Comment:
I'd rather leave this as a separate if statement. All the other parts of the if-elseif are checking the tag except this one. Suggestion:
```
if exp._type then
    -- This expression was already type-checked before. Probably because of
    -- expand_function_returns.
    return exp
end

local tag = exp._tag
if     tag == "ast.Exp.Nil" then
```